ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,439 @@
+# Ruby - Phase 1
+
+The Ruby SDK will be implemented in two loosely defined phases:
+
+* Phase 1 - Initial project, activity worker, client, and test framework
+* Phase 2 - Workflow worker and replayer
+
+This proposal is for phase 1. While proposals and SDK may be done in large phases, work is actually incremental.
+
+Notes for this proposal:
+
+* Everything is subject to change both during proposal time and afterwards during implementation.
+* For each place a decision is made, a "💭 Why X? ..." aside is present explaining why.
+* For each place an open question exists, a "❓ X?" exists to ask it.
+
+## Overview
+
+The Ruby SDK will be backed by Rust [sdk-core](https://github.com/temporalio/sdk-core) and will follow the model set by
+other SDKs.
+
+Much of the work for this proposal and the implementation comes the past proposals in [previous](previous)
+and the implementation built from that. 💭 Why start a new proposal/implementation instead of adapting that? The
+constraints, approach, and requirements changed enough to justify a clean slate and while many things may be leveraged,
+it never went beyond alpha anyways.
+
+Lots of approaches taken here are similar to recent Python and .NET SDKs.
+
+### Requirements
+
+These are the general requirements of the project
+
+* Ruby >= 3.1
+  * 💭 Why not older versions? They are all EOL'd and we need Ractors and other things that come with 3.x.
+* Limited dependencies
+  * This means we will try to make `google-protobuf` the only dependency. We should not depend on `async` or anything
+    else if we can help it.
+  * 💭 Why? Our dependencies as a library have transitive effects on our users, so we have a responsibility to be
+    lightweight.
+* Must be both thread-blocking/thread-safe and async-capable
+  * PoC shows that this is most easily done with Rust block/callback to `Queue` `push`/`pop` since in modern Ruby
+    `Queue.pop` does not block fibers and many other things don't either. Therefore many SDK features like clients and
+    codecs don't have to distinguish between sync/async.
+  * Activities need to work in both ways and _do_ have to make a distinction (discussed later).
+  * 💭 Why async-capable? Core and all SDKs are very parallel and so we want to give that benefit to users if they want
+    it.
+  * 💭 Why thread-blocking? That's the most common way Ruby is used.
```

Comment:
Context related to this topic, ruby @ Stripe was single-threaded by default.  Multithreaded carve-outs did happen but were relatively unusual.
This meant some libraries weren't designed to be thread-safe and used things like class variables.
Default was therefore to have many workflow pollers but only one activity poller per process, which was inefficient, but we added a config flag folks could turn on to make their workers run multiple activity pollers.  I really wanted to make the multi-poller version the default but didn't do so before I left.

Reply:
This is why we plan to also support single-threaded/coroutine-based implementations. It's important to us at Temporal that people be able to do a lot of concurrent things with the SDK in just a single thread. Granted the current design is queue-pop-based which makes it work with fiber-based async. I wasn't planning on supporting other async approaches (e.g. event machine) but if there is enough desire for pluggable-non-fiber-based async, we can look into that.

> Multithreaded carve-outs did happen but were relatively unusual.

These carve-outs were likely required for to use the existing Ruby SDKs.

Reply:
On the carve-outs, I'm referring to the "pre-temporal world."  Later, with the old SDK, we simply limited to one activity poller, which fixed any race conditions.  It was a cautious approach, as most customers probably would have been fine, but there was FUD.